src/ScottPlot5/ScottPlot5/DataSourceUtilities.cs

```diff
@@ -0,0 +1,475 @@
+        public int Compare(double a, double b) => a.CompareTo(b);
+        public int Compare(RootedPixelVector x, RootedPixelVector y) => x.Point.X.CompareTo(y.Point.X);
+        public int Compare(RootedCoordinateVector x, RootedCoordinateVector y) => x.Point.X.CompareTo(y.Point.X);
+
+    }
+
+    public sealed class GenericComparer<T> : IComparer<T>
+    {
+        ///<summary>
+        ///For standard numerics (single, long, int, etc) this will be '<see cref="Comparer{T}.Default"/>', which is significantly faster than this interface implementation
+        ///</summary> 
+        public static readonly IComparer<T> Instance = GetComparer(); 
+
+#pragma warning disable CS8767 // handled by null check
+        public int Compare(T a, T b)
+        {
+            return true switch
+            {
+                true when a is null | b is null => throw new Exception("Generic Comparer can not handle null values"),
+                true when a is Coordinates c && b is Coordinates d => c.X.CompareTo(d.X),
+                true when a is RootedPixelVector x && b is RootedPixelVector y => x.Point.X.CompareTo(y.Point.X),
+                true when a is RootedCoordinateVector x && b is RootedCoordinateVector y => x.Point.X.CompareTo(y.Point.X),
+                // in practice, these should never be used if the 'Instance' field is refer
+                IComparable<T> c => c.CompareTo(b),
+                _ => NumericConversion.GenericToDouble(ref a).CompareTo(NumericConversion.GenericToDouble(ref b))
+            };
+        }
+#pragma warning restore CS8767 // Nullability of reference types in type of parameter doesn't match implicitly implemented member (possibly because of nullability attributes).
+
+        public static IComparer<T> GetComparer()
+        {
+            return Instance ?? default(T) switch
+            {
+                null => throw new ArgumentException($"T should be a struct (non-nullable)"),
+                Coordinates => new GenericComparer<T>(),
+                RootedPixelVector => new GenericComparer<T>(),
+                RootedCoordinateVector => new GenericComparer<T>(),
+                IComparable<T> => Comparer<T>.Default,
+                _ => new GenericComparer<T>()
+            };
+        }
+    }
+
+    public static class DataSourceUtilities
+    {
+
+        /// <summary>
+        /// Check if a collection is in ascending order
+        /// </summary>
+        /// <typeparam name="T">the type to evaluate</typeparam>
+        /// <param name="values">The values</param>
+        /// <param name="comparer">The comparer to use. If not specified, uses Comparer{T}.Default (which is preferred)</param>
+        /// <returns>True if the collection is ascending (and can therefore be used with BinarySearch and GetClosest). Otherwise false.</returns>
+        public static bool IsAscending<T>(this IEnumerable<T> values, IComparer<T>? comparer)
+        {
+            bool isAscending;
```

Comment:
Gotcha. Only suggestion to your improvement is that I used `>=` because binary search still works fine with identical consecutive values, they just can't go backwards. It's more forgiving for scatter plot, which is primary use case currently.

Reply:
It seems to be correct here, the check fails if the previous element is strictly greater than the current one. Equality is allowed. I just inverted your comparison along with the behavior.